crates/ecs/src/systems/mod.rs

```diff
@@ -211,31 +212,44 @@ pub trait SystemParameters: Send + Sync {
     fn component_accesses() -> Vec<ComponentAccessDescriptor>;
 }
 
-/// Something that can be passed to a `ecs::System`.
-pub trait SystemParameter: Send + Sync + Sized {
+/// Something that can be passed to a [`System`].
+pub(crate) trait SystemParameter: Send + Sync + Sized {
     /// Contains a borrow of components from `ecs::World`.
     type BorrowedData<'components>;
 
-    /// Borrows the collection of components of the given type from `ecs::World`.
-    fn borrow<'a>(
-        world: &'a World,
-        signature: &[TypeId],
-    ) -> SystemParameterResult<Self::BorrowedData<'a>>;
+    /// Borrows the collection of components of the given type from [`World`].
+    fn borrow<'world>(
+        world: &'world World,
+        archetypes: &[ArchetypeIndex],
+    ) -> SystemParameterResult<Self::BorrowedData<'world>>;
 
     /// Fetches the parameter from the borrowed data for a given entity.
     /// # Safety
     /// The returned value is only guaranteed to be valid until BorrowedData is dropped
     unsafe fn fetch_parameter(borrowed: &mut Self::BorrowedData<'_>) -> Option<Option<Self>>;
 
     /// A description of what data is accessed and how (read/write).
-    fn component_access() -> ComponentAccessDescriptor;
-
-    /// Returns the `TypeId` of the borrowed data.
-    fn signature() -> TypeId {
-        match Self::component_access() {
-            ComponentAccessDescriptor::Read(type_id, _)
-            | ComponentAccessDescriptor::Write(type_id, _) => type_id,
-        }
+    fn component_access() -> Option<ComponentAccessDescriptor>;
+
+    /// If the [`SystemParameter`] require that the system iterates over entities.
+    /// The system will only run once if all [`SystemParameter`]'s `iterates_over_entities` is `false`.
+    fn iterates_over_entities() -> bool;
+
+    /// The `base_signature` is used for [`SystemParameter`]s that always require a component on the
+    /// entity. This will be `None` for binary/unary filters.
+    ///
+    /// For example, if `(Read<A>, With<B>, With<C>)` is queried, then the `base_signature` will be
+    /// the [`TypeId`]s of `{A, B, C}`. If instead `(Read<A>, Or<With<B>, With<C>>)` is queried,
+    /// then it will just be the [`TypeId`]s of `{A}`. A set of the archetype indices that includes
+    /// all components of the `base_signature` is created and this set is called the `universe`.
+    /// The queried archetypes is find by taking the intersection of the `universe` and the filtered
+    /// versions of the `universe` using the `filter` function.
```

Comment:
Great, thanks!

```suggestion
    /// The `base_signature` is used for [`SystemParameter`]s that always require a component on the
    /// entity. This will be `None` for binary/unary filters.
    ///
    /// For example, if `(Read<A>, With<B>, With<C>)` is queried, then the `base_signature` will be
    /// the [`TypeId`]s of `{A, B, C}`. If instead `(Read<A>, Or<With<B>, With<C>>)` is queried,
    /// then it will just be the [`TypeId`]s of `{A}`. A set of the archetype indices that includes
    /// all components of the `base_signature` is created and this set is called the `universe`.
    /// The queried archetypes are found by taking the intersection of the `universe` and the filtered
    /// versions of the `universe` using the `filter` function.
```